### PR TITLE
[WIP] feat: add unranked match api

### PR DESCRIPTION
**Pull request by Ming1996**
- Add join , leave , get queue API
- Add test case for queue class

### POST `/match/join`
image
image
image

### POST `/match/leave`
image

### GET `/match/queue`
image

image

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -1,36 +1,37 @@
 // @ts-nocheck
-import './env.js';
```
Comment (vck3000) [inline src/app.ts:2]: What was the purpose of re-ordering these? Could you undo these changes unless there's meaningful changes?

File: src/match/queue.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import { randomUUID } from 'crypto';
+
+export const subscribe = () => {};
```
Comment (vck3000) [inline src/match/queue.ts:3]: Not doing anything?

File: src/match/queue.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+type Item = {
```
Comment (vck3000) [inline src/match/queue.ts:4]: What's this an item of?

File: src/match/queue.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+};
+
+export class Queue {
+  queue: Item[] = [];
+  subscribers: any[] = [];
+  subscribe({
+    onJoin,
+    onLeave,
+  }: {
+    onJoin: (playerID: string) => void;
```
Comment (vck3000) [inline src/match/queue.ts:16]: Let's use `playerId` rather than `playerID`. Makes chaining on extra words nicer.

File: src/match/queue.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+export const subscribe = () => {};
+type Item = {
+  id: string;
+  joinAt: number;
+};
+
+export class Queue {
+  queue: Item[] = [];
+  subscribers: any[] = [];
```
Comment (vck3000) [inline src/match/queue.ts:11]: Let's not have `any` here if we can!

File: src/match/queue.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  subscribe({
+    onJoin,
+    onLeave,
+  }: {
+    onJoin: (playerID: string) => void;
+    onLeave: (playerID: string) => void;
+  }) {
+    const id = randomUUID();
+    this.subscribers.push({ id, onJoin, onLeave });
```
Comment (vck3000) [inline src/match/queue.ts:20]: Hmm, two things here:

1) I'd prefer if we don't use a randomUUID for this. Let's just push on a lowercased username here. It'll be easier to debug and manage going forward. If you're really pedantic about using a globally unique-checked piece of data, we should use the user's uid in the mongo model. 

2) Do we need onJoin and onLeave? What's it for?

File: src/match/queue.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
+import { randomUUID } from 'crypto';
+
+export const subscribe = () => {};
+type Item = {
+  id: string;
+  joinAt: number;
+};
+
+export class Queue {
```
Comment (vck3000) [inline src/match/queue.ts:9]: Let's call it `MatchMakingQueue` or something more descriptive. It isn't just any generic Queue that we can use for normal data structure stuff!

File: src/match/queue.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
+import { randomUUID } from 'crypto';
+
+export const subscribe = () => {};
+type Item = {
+  id: string;
+  joinAt: number;
```
Comment (vck3000) [inline src/match/queue.ts:6]: Hmm, is there a typescript type for Date objects?

File: src/match/queue.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+  }
+  getFirstNItems(n: number) {
+    if (n > this.queue.length) {
+      return;
+    }
+    return this.queue.slice(0, n);
+  }
+  deleteMatchedPlayers(playerIDs: string[]) {
+    this.queue = this.queue.filter(({ id }) => !playerIDs.includes(id));
+  }
```
Comment (vck3000) [inline src/match/queue.ts:47]: So for these last three methods `get`, `getFirstNItems` and `deleteMatchedPlayers`, I think it'd be better to not expose these publicly. 

I also think the following would be a better API:

- When the caller code adds a player to the queue, it calls `join()`. 
- `join()` then returns either a null object, or a list containing the usernames that are matched.  
- When `join()` does this, it'll automatically remove those usernames from the list. This way the caller doesn't have to separately call `deleteMatchedPlayers()`. 
- When a player leaves the queue, simply call `leave()`. 

This will simplify our API and make it easier to use :). Later on when we need to matchmake with elo, we'll pass in the user's elo into the `join()` method and add that extra complexity in `join()`.

File: src/routes/match.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+const unrankedQueue = new Queue();
+// join ranked queue API
+router.post('/join', async (req, res) => {
+  const { username }: { username: string } = req.body;
+  if (!username) {
+    return res.status(400).send({ message: 'cad request' });
+  }
+  const usernameLower = username.toLowerCase();
+  const isPlayerExist = await User.find({
```
Comment (vck3000) [inline src/routes/match.ts:14]: Can do `findOne()` here and simply check `if(user)`

`const user = await User.findOne({})`.

File: src/routes/match.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+// get queue or get first n's players
+router.get('/queue', (req, res) => {
+  const { quantity } = req.body;
+  const result = isNaN(quantity)
+    ? unrankedQueue.get()
+    : unrankedQueue.getFirstNItems(quantity);
+  return res.status(200).send({ result });
+});
+
+export default router;
```
Comment (vck3000) [inline src/routes/match.ts:57]: I like the direction you're going here. I think this would be more suitable inside the context of sockets rather than a HTTP endpoint. 

1) HTTP endpoints are single directional. We can't push back information back to the client from here. E.g. when a match is found, it'll be difficult to push that to all users to prompt them to join the room. Ideally we have a live 2 way connection. This is what sockets are best for. 

2) All the logic relating to room management are in the sockets file. Easier to have those together there. 

3) Because sockets are stateful, we already have the username, usernameLower, and even the `User` object itself! It's attached to the socket object.

File: src/match/queue.test.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  });
+
+  it('should return first 2 members', () => {
+    queue.join('test2');
+    queue.join('test3');
+    const result = queue.getFirstNItems(2);
+    const date1 = result[0]?.joinAt;
+    const date2 = result[0]?.joinAt;
+    expect(result.length).toBe(2);
+    expect(result).toEqual([
```
Comment (vck3000) [inline src/match/queue.test.ts:27]: Have some newlines in between blocks of code :)

File: src/match/queue.test.ts [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+      },
+    ]);
+  });
+
+  it('should return first 2 members', () => {
+    queue.join('test2');
+    queue.join('test3');
+    const result = queue.getFirstNItems(2);
+    const date1 = result[0]?.joinAt;
+    const date2 = result[0]?.joinAt;
```
Comment (vck3000) [inline src/match/queue.test.ts:25]: I think we can leave out the `joinAt` field. It won't be used in the logic to determine who will be matched, and if we need to add it in, we can add it in later. Keeping `joinAt` here will contribute to more code pollution and extra stuff in these unit tests that aren't actually being used.

**Review by vck3000 — CHANGES_REQUESTED** [review]

File: src/app.ts (hunk 1 above)
Comment (Ming1996) [reply src/app.ts:2]: oh , I like to press `option+shift+o`(Mac) after coding, vscode will order the import and clean up un-used import automatically

but I can see what you mean, its quite messy in the code change since I just add one line , I can undo them :D

File: src/match/queue.ts (hunk 8 above)
Comment (Ming1996) [reply src/match/queue.ts:6]: joinAt is a timestamp , so I use number here

File: src/match/queue.ts (hunk 6 above)
Comment (Ming1996) [reply src/match/queue.ts:20]: `onJoin`:  call back function when player join the queue - eg: when a user join a queue, use onJoin to call join lobby api
`onLeave`: call back function when player leave the queue
(I add these into comment )

File: src/match/queue.ts (hunk 8 above)
Comment (vck3000) [reply src/match/queue.ts:6]: I mean this:

image

https://www.typescriptlang.org/docs/handbook/2/basic-types.html#explicit-types

File: src/match/queue.ts (hunk 9 above)
Comment (Ming1996) [reply src/match/queue.ts:47]: ` join(playerId: string) {`
    `this.queue.push({`
      `id: playerId,`
     ` joinAt: Date.now(),`
    `});`
    `this.subscribers.forEach((subscriber) => subscriber.onJoin(playerId));`
    `return this.queue`
  }`

Is this what you mean? every time they join, it will return the queue as well

File: src/match/queue.test.ts (hunk 13 above)
Comment (Ming1996) [reply src/match/queue.test.ts:25]: but in the **ranked** match making, we need to consider the join time and expend the range of Elo score, is that right?

Comment (vck3000) [reply src/match/queue.test.ts:26]: We'll leave that for future work :). For now let's not worry about this!

File: src/match/queue.ts (hunk 9 above)
Comment (vck3000) [reply src/match/queue.ts]: Nah, I was thinking it'll only return 6 usernames if there are 6 people in the queue now. If not, then return null.

**Comment by FakhruddinNalawala** [comment]
Deprecated - check PR 530
